File: 53d40c1e2f13e331fc000c26.rb
def fib(n)
  a = 0
  b = 1
  for d in n.abs.digits(2)[(-2..).step(-1)] do
    a, b = [a, 2 * b].map { |x| x * a + b * b }
    a, b = b, a + b if d > 0
  end
  b * [0, 1, (-1) ** n.succ][n <=> 0]
end

File: 53d40c1e2f13e331fc000c26.py
def fib(n):
    a = 0
    b = 1
    for d in bin(abs(n))[3:]:
        a, b = (x * a + b * b for x in (a, 2 * b))
        if d > '0': a, b = b, a + b
    return b if n > 0 else n and -b * (-1) ** -n
